core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -517,59 +530,62 @@ private void innerIndex(Index index) throws IOException {
             }
             final long expectedVersion = index.version();
             if (checkVersionConflict(index, currentVersion, expectedVersion, deleted)) {
-                index.setCreated(false);
-                return;
-            }
-            final long updatedVersion = updateVersion(index, currentVersion, expectedVersion);
-            index.setCreated(deleted);
-            if (currentVersion == Versions.NOT_FOUND && forceUpdateDocument == false) {
-                // document does not exists, we can optimize for create
-                index(index, indexWriter);
+                // skip index operation because of version conflict on recovery
+                return new IndexResult(expectedVersion, false, index.estimatedSizeInBytes());
```

Comment:
PS: for **another** change - I don't like the inconsistencies in how we handle version conflicts during recovery and how we handle them for replicas. For the first we ignore here. For the second we ignore the replication action universe. We should have one way for both.